File: project_euler_unified/problem_952/solution.py
"""
Problem 952: Integer Points on Ellipses

Count integer (x, y) with 3x^2 + 5y^2 <= N, where N = 10^6.

The lattice point count for ellipse Ax^2 + By^2 <= N is approximately
pi*N/sqrt(AB), with error O(sqrt(N)). We compute exactly by slicing:
for each valid x, count y-values satisfying y^2 <= (N - 3x^2) / 5.

Complexity: O(sqrt(N)) time, O(1) space.
"""

import numpy as np
from math import isqrt

def solve(N: int = 10**6) -> int:
    """Count integer points (x, y) with 3x^2 + 5y^2 <= N."""
    count = 0
    max_x = isqrt(N // 3)
    for x in range(-max_x, max_x + 1):
        rem = N - 3 * x * x
        if rem < 0:
            continue
        max_y = isqrt(rem // 5)
        # Verify floor correctness
        while 5 * max_y * max_y > rem:
            max_y -= 1
        count += 2 * max_y + 1
    return count

# --- Compute answer ---
N = 10**6
answer = solve(N)
area_approx = np.pi * N / np.sqrt(15)
print(f"Exact count:     {answer}")
print(f"Area estimate:   {area_approx:.1f}")
print(f"Relative error:  {abs(answer - area_approx) / area_approx:.4%}")
print(answer)

File: project_euler_unified/problem_952/solution.cpp
/*
 * Problem 952: Integer Points on Ellipses
 *
 * Count integer (x, y) with 3x^2 + 5y^2 <= N, N = 10^6.
 *
 * For each x in [-sqrt(N/3), sqrt(N/3)], count y in [-M, M] where
 * M = floor(sqrt((N - 3x^2) / 5)).
 *
 * Complexity: O(sqrt(N)) time, O(1) space.
 */

#include <bits/stdc++.h>
using namespace std;

int main() {
    const long long N = 1000000;
    long long count = 0;

    int max_x = (int)sqrt((double)N / 3.0);
    // Adjust for floating-point imprecision
    while (3LL * (max_x + 1) * (max_x + 1) <= N) max_x++;
    while (3LL * max_x * max_x > N) max_x--;

    for (int x = -max_x; x <= max_x; x++) {
        long long rem = N - 3LL * x * x;
        if (rem < 0) continue;

        int max_y = (int)sqrt((double)rem / 5.0);
        // Correct floating-point floor
        while (5LL * (max_y + 1) * (max_y + 1) <= rem) max_y++;
        while (5LL * max_y * max_y > rem) max_y--;

        count += 2 * max_y + 1;
    }

    cout << count << endl;
    return 0;
}
